src/main/java/com/nvidia/spark/rapids/jni/HyperLogLogPlusPlusHostUDF.java

```diff
@@ -0,0 +1,142 @@
+   */
+  public enum AggregationType {
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge all the sketches into one sketch, output
+     * is a struct scalar with multiple long values.
+     */
+    Reduction(0),
+
+    /**
+     * Merge all HyperLogLogPlusPlus(HLLPP) sketches in the input column into one
+     * sketch. Input is a struct column with multiple long columns which is
+     * consistent with Spark. Output is a struct scalar with multiple long values.
+     */
+    ReductionMERGE(1),
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge the sketches in the same group. Output is
+     * a struct column with multiple long columns which is consistent with Spark.
+     */
+    GroupBy(2),
+
+    /**
+     * Merge HyperLogLogPlusPlus(HLLPP) sketches in the same group.
+     * Input is a struct column with multiple long columns which is consistent with
+     * Spark.
+     */
+    GroupByMerge(3);
+
+    final int nativeId;
+
+    AggregationType(int nativeId) {
+      this.nativeId = nativeId;
+    }
+  }
+
+  /**
+   * Create a HyperLogLogPlusPlus(HLLPP) host UDF
+   */
+  private static long createHLLPPHostUDF(AggregationType type, int precision) {
+    return createHLLPPHostUDF(type.nativeId, precision);
+  }
+
+  /**
+   * Compute the approximate count distinct value from sketch values.
+   *
+   * The input is sketch values, must be given in the format:
+   * `Struct<INT64, INT64,...>`,
+   * The value of num_registers_per_sketch = 2^precision
+   * The children num of this Struct is: num_registers_per_sketch / 10 + 1,
+   * Here 10 means a INT64 contains 10 register values,
+   * each register value is 6 bits.
+   * Register value is the number of leading zero bits in xxhash64 hash code.
+   * xxhash64 hash code is 64 bits, Register value is 6 bits,
+   * 6 bits is enough to hold the max value 64.
+   *
+   * @param input     The sketch column which constains Struct<INT64, INT64, ...>
+   *                  values.
+   * @param precision The num of bits for HLLPP register addressing.
+   * @return A INT64 column with each value indicates the approximate count
+   *         distinct value.
+   */
+  public static ColumnVector estimateDistinctValueFromSketches(ColumnView input, int precision) {
+    return new ColumnVector(estimateDistinctValueFromSketches(input.getNativeView(), precision));
+  }
+
+  private static native long createHLLPPHostUDF(int type, int precision);
+
+  private static native long estimateDistinctValueFromSketches(long inputHandle, int precision);
+
+  private AggregationType type;
+  private int precision;
+
+  /**
+   * TODO: move this to cuDF HostUDFWrapper
+   */
+  @Override
+  public void close() throws Exception {
+    close(udfNativeHandle);
+  }
+
+  /**
+   * TODO: move this to cuDF HostUDFWrapper
+   */
+  static native void close(long ptr);
```

Comment:
Oh moving these function to cudf JNI is possible.

Reply:
OK to me, but I have concerns:
If we have more derived classes in future, every derived classes should have a JNI close implemetation.
`udfNativeHandle` is in  `HostUDFWrapper`, seems it's reasonable to close `udfNativeHandle`  in `HostUDFWrapper`. But udfNativeHandle is created in the derived class. I'm not sure which place is better.

Reply:
We have some options:
 - If we close `udfNativeHandle` in the base class `HostUDFWrapper` in some method like `closeHandle()`, the derived class still needs to have a `close()` method to close its resources. This `close()` method will call `closeHandle()` in the base class.
 - If we close `udfNativeHandle` in the derived class through its `close()`, every derived class will have to do this same operation. Since every derived class has to implement a native method to create its native handle anyway. So each UDF has a pair of `create()` and `close()`.
 - We will implement `close()` in the base class `HostUDFWrapper` and the derived Java class will override it only if it has more resources to close. This seems the best option. @res-life do you have time to refactor that, after this PR?

Reply:
Yes, I'll refactor that.
I prefer the third one: close the `udfNativeHandle` in in the base class HostUDFWrapper. And one little improvement:

```java
class HostUDFWrapper {
   // derived classes should override this function
   // derived classes MUST close the resources they created in here.
   abstract void closeMoreResource() {
   }

   void close() {
       // close the udfNativeHandle first,
       // then close the other resources created in the  derived classes
       closeMoreResource();
   }
}

```

Reply:
done: https://github.com/rapidsai/cudf/pull/17770